framework/src/main/java/org/checkerframework/framework/source/CompositeChecker.java

```diff
@@ -0,0 +1,756 @@
+package org.checkerframework.framework.source;
+
+import com.google.common.collect.ImmutableSet;
+import com.sun.source.tree.CompilationUnitTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.util.TreePath;
+import com.sun.tools.javac.processing.JavacProcessingEnvironment;
+import com.sun.tools.javac.util.Context;
+import com.sun.tools.javac.util.Log;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeSet;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.TypeElement;
+import javax.tools.Diagnostic;
+import org.checkerframework.checker.interning.qual.FindDistinct;
+import org.checkerframework.checker.interning.qual.InternedDistinct;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.checker.signature.qual.ClassGetName;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.reflection.MethodValChecker;
+import org.checkerframework.framework.util.TreePathCacher;
+import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.InternalUtils;
+import org.checkerframework.javacutil.TypeSystemError;
+import org.checkerframework.javacutil.UserError;
+import org.plumelib.util.ArrayMap;
+import org.plumelib.util.CollectionsPlume;
+import org.plumelib.util.StringsPlume;
+
+/**
+ * An abstract {@link SourceChecker} that provides support for subcheckers and manages their
+ * messages. It doesn't require an extending checker to provide a qualifier hierarchy, type factory
+ * or visitor. Hence, it can be used similar to an AggregateChecker, with the additional service of
+ * managing messages.
+ */
+public abstract class CompositeChecker extends SourceChecker {
+
+  /** Create a new CompositeChecker. */
+  protected CompositeChecker() {}
+
+  @Override
+  public void initChecker() {
+    // initialize all checkers and share options as necessary
+    for (BaseTypeChecker checker : getSubcheckers()) {
```

Comment:
It's kind of unusual for a class to reference its own subclass, as a general design rule. So, I'd like to be sure that there's not an alternative here that allows `CompositeChecker` to avoid needing to reference `BaseTypeChecker` at all.

Reply:
I wanted to change as little as possible for the BaseTypeChecker. Composite checker is subclassed directly exclusively by the refactored RLC. However, many classes with subcheckers subclass BaseTypeChecker directly. They want their subcheckers to be BaseTypeCheckers. Would the subcheckers be CompositeCheckers, they would have to cast them to access their typefactories and analyses.

I know, it's a bit unfortunate, but I didn't find a better solution.